Print warning if flutter drive is run in debug

liyuqian opened the pull request:
## Description

Print actionable warnings if `flutter drive` (where most of our performance benchmarks come from) is run in debug mode and it tries to gather benchmarks using `traceAction`.

## Related Issues

https://github.com/flutter/flutter/issues/30625

## Tests

I added the following tests:
* drive_perf_debug_warning devicelab test

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

liyuqian committed:
Print warning if flutter drive is run in debug

jonahwilliams commented:
`flutter drive` is also used for integration tests, not just benchmarks. It seems like this is going to be a lot of extra noise in that case

Hixie commented:
We should only show this for benchmarks.

FWIW, here's the code for flutter_test that does the equivalent:
https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/widget_tester.dart#L156

Important things to notice:
* it only runs for benchmarks
* it only shows the message if the benchmark is run in debug mode
* the message is pretty and uses fancy Unicode

liyuqian committed:
Only print warnings if traceAction is called

dnfield commented on packages/flutter_driver/lib/src/driver/driver.dart:
```diff
@@ -764,6 +780,16 @@ class FlutterDriver {
   }
 
+  Future<bool> _isPrecompiledMode() async {
+    final List<Map<String, dynamic>> flags = await getVmFlags();
```
How long does this take?  Would it be faster to do soemthing like:

```
bool assertsEnabled = false;
assert(() {
  assertsEnabled = true;
}());
return assertsEnabled;
```
?

liyuqian replied:
I tried that before exploring and fixing this buggy `getVmFlags` function (which was never used in our repo). Somehow, the code inside `assert` is never run. I thought maybe `flutter drive` itself is running the host dart code in non-debug mode, while the client dart code is running in debug mode.

liyuqian replied:
BTW, since this is run after the tracing is done, it shouldn't affect the tracing numbers. In terms of the test time cost, it probably adds no more than 0.5 second since I couldn't see any significant test slowdown.

dnfield reviewed (approved):
LGTM

liyuqian committed:
Add a devicelab test for the warning message

This test will also make sure that the VM flags are working as intended.
(Previously, `getVmFlags` is unused in our repo and will throw exception
upon calling.)

liyuqian committed:
Handle null flags

liyuqian commented:
@Hixie @jonahwilliams : how about the new version which only prints warning if `traceAction` is called during `flutter drive`. I've added a devicelab test to test both this warning, and the previously unused `getVmFlags` function.

Hixie commented on packages/flutter_driver/lib/src/driver/driver.dart:
```diff
@@ -708,7 +724,7 @@ class FlutterDriver {
   Future<List<Map<String, dynamic>>> getVmFlags() async {
     final Map<String, dynamic> result = await _peer.sendRequest('getFlagList');
-    return result['flags'];
+    return result != null ? result['flags'].cast<Map<String,dynamic>>() : <Map<String, dynamic>>[];
```
can the fallback be const?

liyuqian replied:
Yes. Done.

Hixie commented on packages/flutter_driver/lib/src/driver/driver.dart:
```diff
@@ -788,6 +814,10 @@ class FlutterDriver {
     }
```
update the documentation here as well

liyuqian replied:
Done.

Hixie commented on dev/devicelab/bin/tasks/drive_perf_debug_warning.dart:
```diff
@@ -0,0 +1,38 @@
+      'Could not find the following warning message piece: $warningPiece'
+    );
+  }
+}
```
you should also test the reverse (that running with --profile doesn't show the warning).

liyuqian replied:
Done.

Hixie commented:
Approach seems reasonable.

liyuqian committed:
Make empty list const

liyuqian committed:
Add test for profile mode too

jonahwilliams reviewed (approved):
LGTM!
